<a href="https://colab.research.google.com/github/mjiii25/prac_class/blob/main/HW2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Neural Network**

In [4]:
import tensorflow
tensorflow.keras.__version__

'2.4.0'

In [5]:
from tensorflow.keras.datasets import mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

11493376/11490434 [==============================] - 0s 0us/step


- train_images, train_labels 
: training set (the data that the model will learn from)

- test_images, test_labels : test set (the data that will be tested)

##Training data

In [6]:
train_images.shape

(60000, 28, 28)

In [7]:
len(train_labels)

60000

In [8]:
train_labels

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

## Test data


In [9]:
test_images.shape

(10000, 28, 28)

In [10]:
len(test_labels)

10000

In [11]:
test_labels

array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

###_workflow_

1. present neural network with the training data
2. network learns to associate images and labels.
3. ask the network to produce predictions for test data.
4. verify if these predictions match the labels from test data.

##Building our network

In [12]:
from tensorflow.keras import models
from tensorflow.keras import layers

network = models.Sequential()
network.add(layers.Dense(512, activation = 'relu', input_shape = (28*28,)))
network.add(layers.Dense(10, activation = 'softmax'))

- layer : data-processing module working as a **_filter_** for data


- [ our data ] two densely-connected neural layers
  (1) relu (2) softmax

## Getting ready for training 

In [13]:
network.compile ( optimizer = 'rmsprop',
                loss = 'categorical_crossentropy',
                metrics = ['accuracy'])

1. Reshape into the shape that the network expects.

2. Perform scaling so that all values are in the [0,1].




In [14]:
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255

Categorically encode the labels.

In [15]:
from tensorflow.keras.utils import to_categorical

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

## Fit the Model to the Training Data

In [16]:
network.fit( train_images, train_labels, epochs = 5, batch_size = 128)

Epoch 1/5
469/469 [==============================] - 4s 3ms/step - loss: 0.4257 - accuracy: 0.8760
Epoch 2/5
469/469 [==============================] - 1s 3ms/step - loss: 0.1133 - accuracy: 0.9673
Epoch 3/5
469/469 [==============================] - 1s 3ms/step - loss: 0.0679 - accuracy: 0.9794
Epoch 4/5
469/469 [==============================] - 1s 3ms/step - loss: 0.0478 - accuracy: 0.9855
Epoch 5/5
469/469 [==============================] - 1s 3ms/step - loss: 0.0374 - accuracy: 0.9889


: test accuracy of training set = **98.9%**

## Checking How Much Our Model Performs Well on the Test Set




In [17]:
test_loss, test_acc = network.evaluate( test_images, test_labels)

313/313 [==============================] - 1s 2ms/step - loss: 0.0716 - accuracy: 0.9780


In [18]:
print('test_acc:', test_acc)

test_acc: 0.9779999852180481


: test accuracy of test set = **97.8%**

## _Conclusion_

Test accuracy of the test set is 97.8%, which is lower than that of training set. 

###→ **overfitting**

###: machine learing model seems to perform worse on new data than on previous data 
